cdap-security-spi/src/main/java/io/cdap/cdap/security/spi/authorization/AccessEnforcer.java

```diff
@@ -0,0 +1,85 @@
+ * Enforces authorization for a {@link Principal} to perform an {@link Permission} on an {@link EntityId}.
+ */
+@Beta
+public interface AccessEnforcer {
+
+  /**
+   * Enforces authorization for the specified {@link Principal} for the specified {@link Permission} on the specified
+   * {@link EntityId}.
+   *
+   * @param entity the {@link EntityId} on which authorization is to be enforced
+   * @param principal the {@link Principal} that performs the permission
+   * @param permission the {@link Permission} being performed
+   * @throws UnauthorizedException if the principal is not authorized to perform the specified permission on the entity
+   */
+  default void enforce(EntityId entity, Principal principal, Permission permission) throws AccessException {
+    enforce(entity, principal, Collections.singleton(permission));
+  }
+
+  /**
+   * Enforces authorization for the specified {@link Principal} for the specified {@link Permission permissions} on the
+   * specified {@link EntityId}.
+   *
+   * @param entity the {@link EntityId} on which authorization is to be enforced
+   * @param principal the {@link Principal} that performs the permissions
+   * @param permissions the {@link Permission permissions} being performed
+   * @throws UnauthorizedException if the principal is not authorized to perform the specified permissions on the entity
+   */
+  void enforce(EntityId entity, Principal principal, Set<? extends Permission> permissions) throws AccessException;
+
+  /**
+   * Checks whether a single {@link EntityId} is visible to the specified {@link Principal}.
+   * An entity is visible to a principal if the principal has {@link io.cdap.cdap.proto.security.StandardPermission#GET}
+   * priviledge on the entity.
+   * However, visibility check behavior can be overwritten at the authorization extension level.
+   *
+   * @param entityId the entity on which the visibility check is to be performed
+   * @param principal the principal to check the visibility for
+   * @throws UnauthorizedException if the entity is not visible to the principal
+   */
+  default void isVisible(EntityId entityId, Principal principal) throws AccessException {
+    if (isVisible(Collections.singleton(entityId), principal).isEmpty()) {
+      throw new UnauthorizedException(principal, entityId);
+    }
+  }
+
+  /**
+   * Checks whether the set of {@link EntityId}s are visible to the specified {@link Principal}.
+   * An entity is visible to a principal if the principal has any privileges on the entity, or any of its descendants.
+   * However, visibility check behavior can be overwritten at the authorization extension level.
+   *
+   * @param entityIds the entities on which the visibility check is to be performed
+   * @param principal the principal to check the visibility for
+   * @return a set of entities that are visible to the principal
+   */
+  Set<? extends EntityId> isVisible(Set<? extends EntityId> entityIds, Principal principal) throws AccessException;
```

Comment:
Since we're already changing the SPI here, can we simply remove the isVisible definitions and instead make them enforce on StandardPermission.GET from the caller? We'll need to add a new enforce() method which takes in a set of entities and set of permissions and returns a subset of entities which the caller has access to. This would help us unify the SPI to all use permissions instead of having a separate set of methods to enforce visibility.

This also may require some refactoring on the caller side too because, from my initial understanding, isVisible is meant to also allow you to see an entity if you have permissions on the entity or any of its parents, but this is not codified into the interface implementation in CDAP and delegated to the authorization extension which represents an interface inconsistency.

Reply:
On a second thought, this may not be an easy switch to make initially and may require some thinking and/or refactoring.

Reply:
Yep, we can do it in the follow-up PR